porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/packagevariantset_controller.go

```diff
@@ -0,0 +1,533 @@
+		Revision: upstream.Revision,
+	}, nil
+
+}
+
+func (r *PackageVariantSetReconciler) unrollDownstreamTargets(ctx context.Context,
+	pvs *api.PackageVariantSet,
+	upstreamPackageName string,
+	repoList *configapi.RepositoryList) ([]*pkgvarapi.Downstream, error) {
+	var result []*pkgvarapi.Downstream
+	for _, target := range pvs.Spec.Targets {
+		switch {
+		case target.Package != nil:
+			// an explicit repo/package name pair
+			result = append(result, r.repoPackagePair(&target, upstreamPackageName))
+
+		case target.Repositories != nil:
+			// a label selector against a set of repositories
+			pkgs, err := r.repositorySet(&target, upstreamPackageName, repoList)
+			if err != nil {
+				return nil, fmt.Errorf("error when selecting repository set: %v", err)
+			}
+			result = append(result, pkgs...)
+
+		case target.Objects != nil:
+			// a selector against a set of arbitrary objects
+			pkgs, err := r.objectSet(ctx, &target, upstreamPackageName)
+			if err != nil {
+				return nil, fmt.Errorf("error when selecting object set: %v", err)
+			}
+			result = append(result, pkgs...)
+		}
+	}
+	return result, nil
+}
+
+func (r *PackageVariantSetReconciler) repoPackagePair(target *api.Target,
+	upstreamPackageName string) *pkgvarapi.Downstream {
+	downstreamPackageName := target.Package.Name
+	if downstreamPackageName == "" {
+		downstreamPackageName = upstreamPackageName
+	}
+	return &pkgvarapi.Downstream{
+		Repo:    target.Package.Repo,
+		Package: downstreamPackageName,
+	}
+}
+
+func (r *PackageVariantSetReconciler) repositorySet(
+	target *api.Target,
+	upstreamPackageName string,
+	repoList *configapi.RepositoryList) ([]*pkgvarapi.Downstream, error) {
+	var result []*pkgvarapi.Downstream
+	for _, repo := range repoList.Items {
+		if labelsMatch(target.Repositories, repo.Labels) {
+			repoAsRNode, err := r.convertObjectToRNode(&repo)
+			if err != nil {
+				return nil, fmt.Errorf("error converting repo to RNode: %v", err)
+			}
+			downstreamPackageName, err := r.getDownstreamPackageName(target.PackageName, upstreamPackageName, repoAsRNode)
+			if err != nil {
+				return nil, err
+			}
+			result = append(result, &pkgvarapi.Downstream{
+				Repo:    repo.Name,
+				Package: downstreamPackageName,
+			})
+		}
+	}
+	return result, nil
+}
+
+func labelsMatch(matchLabels map[string]string, repoLabels map[string]string) bool {
+	for labelKey, labelValue := range matchLabels {
+		if repoValue, labelFound := repoLabels[labelKey]; !labelFound || labelValue != repoValue {
+			return false
+		}
+	}
+	return true
+}
+
+func (r *PackageVariantSetReconciler) objectSet(ctx context.Context,
+	target *api.Target,
+	upstreamPackageName string) ([]*pkgvarapi.Downstream, error) {
+	// TODO: Implement this
+	return nil, fmt.Errorf("specifying a set of objects in the target is not yet supported")
+}
+
+func (r *PackageVariantSetReconciler) getDownstreamPackageName(targetName *api.PackageName,
+	upstreamPackageName string,
+	obj *yaml.RNode) (string, error) {
```

Comment:
I think it might be easier to just use Unstructured here instead of converting to RNode. At least it is easier for future users who might not know kyaml. Fetching fields should be pretty straight-forward using the functions in the unstructured package.

Reply:
I agree with your sentiment and am glad you mentioned it, so I played around with `Unstructured` for a little bit to see if we could use it here instead, but IMO using RNode will actually be easier. From the [unstructured docs](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured):

```
NestedFieldNoCopy returns a reference to a nested field. Returns false if value is not found and an error if unable to traverse obj.

Note: fields passed to this function are treated as keys within the passed object; no array/slice syntax is supported.
```

From what I can tell, `NestedFieldNoCopy` is used under the hood for most, if not all, of the other `NestedXXX` functions in the library, so I ran into issues with trying to support fetching fields from nested array/slice elements. For example, I want to be able to specify a path like `spec.containers[1].image`, but unless I am missing something (which is very possible), the unstructured package doesn't provide a way to do so easily. In general I do agree that kyaml has a steeper learning curve and probably isn't as popular among kubernetes devs as Unstructured, but I think in this specific case, not using kyaml would mean rewriting nontrivial logic to handle array elements, so I feel that using kyaml directly makes more sense.

Reply:
That makes sense. My only question is if we have to specify a path that traverses into a specific entry in an array, should we actually be looking at all entries.

Reply:
Could you elaborate on that? The path specified would come from a user-specified field in `PackageVariantSet` to fetch a particular field, I'm not sure that I understand the scenario where we would need to look at all entries?

Reply:
I might have been confused by the `spec.containers[1].image` example. Usually I get suspicious when there is code checking something on only one container, since there might be more than one and we can't usually assume a specific ordering. I see now that we allow the users to specify a field path. I agree that kyaml is much better than unstructured for this.